src/Microsoft.IdentityModel.Clients.ActiveDirectory/Platforms/iOS/AuthenticationAgentUIViewController.cs

```diff
@@ -152,8 +150,8 @@ public override void ViewDidLoad()
 
         private void CancelAuthentication(object sender, EventArgs e)
         {
-            callbackMethod(new AuthorizationResult(AuthorizationStatus.UserCancel, null));
             this.DismissViewController(true, null);
+            callbackMethod(new AuthorizationResult(AuthorizationStatus.UserCancel, null));
```

Comment:
I would propose to change this line to:

`this.DismissViewController(true, ()=> callbackMethod(new AuthorizationResult(AuthorizationStatus.UserCancel, null)));`

Moving the callback method to the completionHandler would ensure it will not fire until the controller is fully dismissed and removed from the view hierarchy.

The same would apply to the following code snippets:

https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/blob/40e622f5aa279bf7c5852cdca8bd572580039afa/src/Microsoft.IdentityModel.Clients.ActiveDirectory/Platforms/iOS/AuthenticationAgentUIViewController.cs#L96-L97

https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/blob/40e622f5aa279bf7c5852cdca8bd572580039afa/src/Microsoft.IdentityModel.Clients.ActiveDirectory/Platforms/iOS/AuthenticationAgentUIViewController.cs#L126-L127

These would have to be changed to 

`
                    this.DismissViewController(true, ()=>
                        callbackMethod(new AuthorizationResult(AuthorizationStatus.Success, request.Url.ToString())));`

and 

`
                    this.DismissViewController(true, ()=> callbackMethod(result));`

respectively.